Add more client info flags to module API

martinrvisser opened the pull request:
Adding more client flags in the information populated by `ValkeyModule_GetClientInfoById()`:

    VALKEYMODULE_CLIENTINFO_FLAG_PRIMARY
    VALKEYMODULE_CLIENTINFO_FLAG_REPLICA
    VALKEYMODULE_CLIENTINFO_FLAG_MONITOR
    VALKEYMODULE_CLIENTINFO_FLAG_MODULE
    VALKEYMODULE_CLIENTINFO_FLAG_AUTHENTICATED
    VALKEYMODULE_CLIENTINFO_FLAG_EVER_AUTHENTICATED
    VALKEYMODULE_CLIENTINFO_FLAG_FAKE

Closes #2590

martinrvisser committed:
Adding client flags for module usage  (closes 2590)
Signed-off-by: Martin Visser <mvisser@hotmail.com>

Signed-off-by: martinrvisser <mvisser@hotmail.com>

martinrvisser committed:
Adding client flags for module usage  (closes 2590)

Signed-off-by: martinrvisser <mvisser@hotmail.com>

zuiderkwast commented on src/module.c:
```diff
@@ -3826,6 +3826,13 @@ int modulePopulateClientInfoStructure(void *ci, client *client, int structver) {
     if (client->flag.blocked) ci1->flags |= VALKEYMODULE_CLIENTINFO_FLAG_BLOCKED;
     if (client->conn->type == connectionTypeTls()) ci1->flags |= VALKEYMODULE_CLIENTINFO_FLAG_SSL;
     if (client->flag.readonly) ci1->flags |= VALKEYMODULE_CLIENTINFO_FLAG_READONLY;
+    if (client->flag.primary) ci1->flags |= VALKEYMODULE_CLIENTINFO_FLAG_PRIMARY;
```
These flags are documented in the comment before `VM_GetClientInfoById`  which is used in the generated documentation. Please add the new flags there. Let's also mention that these were added in 9.1.0.

martinrvisser replied:
@zuiderkwast added these in the comments in module.c  Where do you want to mention these were added in 9.1?

zuiderkwast replied:
Just below the list of flags? E.g. "The flags VALKEYMODULE_CLIENTINFO_FLAG_PRIMARY and below were added in Valkey 9.1".

martinrvisser replied:
I added this comment in the header file, seemed like the right place for it?

zuiderkwast replied:
I guess many users will read the header file, but it's the comments in module.c that are used for the generated docs at https://valkey.io/topics/modules-api-ref/ so that's why I'm being so picky about these comments. :) There is a script to extract the markdown docs from the .c file at https://github.com/valkey-io/valkey/blob/unstable/utils/generate-module-api-doc.rb (if you want to try it or just FYI).

So let's add the comment in the .c file too, just after the list of flags.

zuiderkwast replied:
The current docs for ValkeyModule_GetClientInfoById: https://valkey.io/topics/modules-api-ref/#ValkeyModule_GetClientInfoById

zuiderkwast reviewed (commented):
> @zuiderkwast this PR is just adding new constants to module API, do we need the core team to approve this change?

Ricardo, technically yes. (But since this is a very small and safe API change, I would take your approval and mine as good enough.)

@valkey-io/core-team Please +1.

martinrvisser committed:
Address PR feedback: documenting additional flags.

Signed-off-by: martinrvisser <mvisser@hotmail.com>

zuiderkwast commented on src/module.c:
```diff
@@ -3886,6 +3893,12 @@ int modulePopulateReplicationInfoStructure(void *ri, int structver) {
  *     VALKEYMODULE_CLIENTINFO_FLAG_UNIXSOCKET   Client using unix domain socket.
  *     VALKEYMODULE_CLIENTINFO_FLAG_MULTI        Client in MULTI state.
  *     VALKEYMODULE_CLIENTINFO_FLAG_READONLY     Client in ReadOnly state.
+ *     VALKEYMODULE_CLIENTINFO_FLAG_PRIMARY      Client is a primary.
```
It's not very clear that this is about replication. Maybe we can improve the description for it. Maybe something like this?

```suggestion
 *     VALKEYMODULE_CLIENTINFO_FLAG_PRIMARY      Client is a fake client used
 *                                               for applying replicated
 *                                               commands from the primary.
```

Btw, perhaps we should make this a proper bullet list rather than a code block...

zuiderkwast commented on src/module.c:
```diff
@@ -3886,6 +3893,12 @@ int modulePopulateReplicationInfoStructure(void *ri, int structver) {
+ *     VALKEYMODULE_CLIENTINFO_FLAG_MONITOR      Client in monitor mode.
+ *     VALKEYMODULE_CLIENTINFO_FLAG_MODULE       Client is a module.
+ *     VALKEYMODULE_CLIENTINFO_FLAG_AUTHENTICATED Client has been authenticated.
+ *     VALKEYMODULE_CLIENTINFO_FLAG_EVER_AUTHENTICATED Client has successfully been authenticated in its lifetime.
```
To align the descriptions with the other lines, it's possible to do like this, only for the long flag names. (That's how it's done in man pages.)

```suggestion
 *     VALKEYMODULE_CLIENTINFO_FLAG_AUTHENTICATED
 *                                               Client has been authenticated.
 *     VALKEYMODULE_CLIENTINFO_FLAG_EVER_AUTHENTICATED
 *                                               Client has successfully been
 *                                               authenticated in its lifetime.
```

zuiderkwast reviewed (commented):
If a module is compiled with the newest valkeymodule.h and is loaded in an older Valkey version, how does the module understand if the server will set these flags or not?

Perhaps we should just document that these new flags are added in 9.1. The module can use `ValkeyModule_GetServerVersion` to find out which version is running.

A have a few more small comments about the API documentation.

martinrvisser committed:
Update src/module.c

Co-authored-by: Viktor Söderqvist <viktor.soderqvist@est.tech>
Signed-off-by: martinrvisser <martinrvisser@users.noreply.github.com>

martinrvisser committed:
Update src/module.c

Co-authored-by: Viktor Söderqvist <viktor.soderqvist@est.tech>
Signed-off-by: martinrvisser <martinrvisser@users.noreply.github.com>

martinrvisser committed:
Fix typo in Valkey module comments

Corrected the comment from 'replic' to 'replica' and added note for ValkeyModuleClientInfo flags.

Signed-off-by: martinrvisser <martinrvisser@users.noreply.github.com>

zuiderkwast commented on src/module.c:
```diff
@@ -3886,6 +3893,17 @@ int modulePopulateReplicationInfoStructure(void *ri, int structver) {
+ *                                               authenticated in its lifetime.
+ *     VALKEYMODULE_CLIENTINFO_FLAG_FAKE         Fake clients are internal to valkey.
  *
  * However passing NULL is a way to just check if the client exists in case
```
```suggestion
 * Note: The flags VALKEYMODULE_CLIENTINFO_FLAG_PRIMARY and below were added in Valkey 9.1.
 *
 * Passing NULL is a way to just check if the client exists in case
```

zuiderkwast committed:
Update src/module.c

Signed-off-by: Viktor Söderqvist <viktor.soderqvist@est.tech>

zuiderkwast commented on src/valkeymodule.h:
```diff
@@ -674,14 +674,22 @@ static const ValkeyModuleEvent ValkeyModuleEvent_ReplicationRoleChanged = {VALKE
 #define _VALKEYMODULE_SUBEVENT_ATOMIC_SLOT_MIGRATION_NEXT 6
 
-/* ValkeyModuleClientInfo flags. */
+/* ValkeyModuleClientInfo flags. 
```
Remove trailing space (clang-format check failed)

```suggestion
/* ValkeyModuleClientInfo flags.
```

zuiderkwast committed:
Remove trailing space detected by clang-format

Signed-off-by: Viktor Söderqvist <viktor.soderqvist@est.tech>

zuiderkwast reviewed (approved):
Sorry for the delay!
